File: src/chapters/least-squares/content/code/line.cpp
#include <iostream>
#include <vector>
using namespace std;
using Vec = vector<double>;

// Least-squares line y = a + b x via the 2x2 normal equations.
void line_fit(const Vec& x, const Vec& y, double& a, double& b) {
    int n = x.size();
    double Sx = 0, Sy = 0, Sxx = 0, Sxy = 0;
    for (int i = 0; i < n; ++i) {
        Sx += x[i]; Sy += y[i];
        Sxx += x[i] * x[i]; Sxy += x[i] * y[i];
    }
    b = (n * Sxy - Sx * Sy) / (n * Sxx - Sx * Sx);   // slope
    a = (Sy - b * Sx) / n;                            // intercept
}

int main() {
    Vec x = {0, 1, 2, 3, 4}, y = {1, 3, 2, 5, 4};
    double a, b;
    line_fit(x, y, a, b);
    cout.precision(4);
    cout << fixed << "slope b = " << b << ", intercept a = " << a << "\n";
}
// -> slope b = 0.8000, intercept a = 1.4000

File: src/chapters/least-squares/content/code/line.js
// Least-squares line y = a + b x via the 2x2 normal equations.
function lineFit(x, y) {
  const n = x.length;
  const Sx = x.reduce((s, xi) => s + xi, 0);
  const Sy = y.reduce((s, yi) => s + yi, 0);
  const Sxx = x.reduce((s, xi) => s + xi * xi, 0);
  const Sxy = x.reduce((s, xi, i) => s + xi * y[i], 0);
  const b = (n * Sxy - Sx * Sy) / (n * Sxx - Sx * Sx); // slope
  const a = (Sy - b * Sx) / n; // intercept
  return { a, b };
}

const x = [0, 1, 2, 3, 4];
const y = [1, 3, 2, 5, 4];
const { a, b } = lineFit(x, y);
console.log("slope b =", b, ", intercept a =", a);
// -> slope b = 0.8, intercept a = 1.4
